src/theory/quantifiers/ho_trigger.cpp

```diff
@@ -0,0 +1,343 @@
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of higher-order trigger class
+ **/
+
+#include <stack>
+
+#include "theory/quantifiers/ho_trigger.h"
+#include "theory/quantifiers/term_database.h"
+#include "theory/quantifiers/term_util.h"
+#include "theory/quantifiers_engine.h"
+#include "theory/theory_engine.h"
+#include "theory/uf/equality_engine.h"
+#include "theory/uf/theory_uf_rewriter.h"
+#include "util/hash.h"
+
+using namespace std;
+using namespace CVC4::kind;
+using namespace CVC4::context;
+
+namespace CVC4 {
+namespace theory {
+namespace inst {
+
+HigherOrderTrigger::HigherOrderTrigger( QuantifiersEngine* qe, Node q, std::vector< Node >& nodes, 
+                                        std::map< Node, std::vector< Node > >& ho_apps ) :
+Trigger( qe, q, nodes ), d_ho_var_apps( ho_apps ) {
+  // process the higher-order variable applications
+  for( std::map< Node, std::vector< Node > >::iterator it = d_ho_var_apps.begin(); 
+       it != d_ho_var_apps.end(); ++it ){
+    Node n = it->first;
+    d_ho_var_list.push_back( n );
+    TypeNode tn = n.getType();
+    Assert( tn.isFunction() );
+    if( Trace.isOn("ho-quant-trigger") ){
+      Trace("ho-quant-trigger") << "  have " << it->second.size();
+      Trace("ho-quant-trigger") << " patterns with variable operator " << n << ":" << std::endl;
+      for( unsigned j=0; j<it->second.size(); j++ ){
+        Trace("ho-quant-trigger") << "  " << it->second[j] << std::endl;
+      }
+    }
+    if( d_ho_var_types.find( tn )==d_ho_var_types.end() ){
+      Trace("ho-quant-trigger") << "  type " << tn << " needs higher-order matching." << std::endl;
+      d_ho_var_types.insert( tn );
+    }
+    // make the bound variable lists
+    d_ho_var_bvl[n] = NodeManager::currentNM()->getBoundVarListForFunctionType( tn );
+    for( unsigned j=0; j<d_ho_var_bvl[n].getNumChildren(); j++ ){
+      d_ho_var_bvs[n].push_back( d_ho_var_bvl[n][j] );
+    }
+  }
+}
+
+HigherOrderTrigger::~HigherOrderTrigger() {
+
+}
+
+void HigherOrderTrigger::collectHoVarApplyTerms( Node q, TNode n, std::map< Node, std::vector< Node > >& apps ) {
+  std::vector< Node > ns;
+  ns.push_back(n);
+  collectHoVarApplyTerms( q, ns, apps );
+}
+
+void HigherOrderTrigger::collectHoVarApplyTerms( Node q, std::vector< Node >& ns, std::map< Node, std::vector< Node > >& apps ) {
+  std::unordered_set<TNode, TNodeHashFunction> visited;
+  // whether the visited node is a child of a HO_APPLY chain
+  std::unordered_map<TNode, bool, TNodeHashFunction> withinApply;
+  std::unordered_set<TNode, TNodeHashFunction>::iterator it;
+  std::stack<TNode> visit;
+  TNode cur;
+  for( unsigned i=0; i<ns.size(); i++ ){
+    visit.push(ns[i]);
+    withinApply[ns[i]] = false;
+    do {
+      cur = visit.top();
+      visit.pop();
+      it = visited.find(cur);
+
+      if (it == visited.end()) {
```

Comment:
Why not skip the iteration if the element is found? The code gets less indented this way. This occurs in plenty of places.